File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/sensors/temperature/TemperatureSensor.java
package io.treehopper.libraries.sensors.temperature;

/**
 * Temperature sensor
 */
public abstract class TemperatureSensor implements Temperature {

    protected double celsius = 0;
    protected boolean autoUpdateWhenPropertyRead = true;

    public static double toKelvin(double celsius) {
        return celsius + 273.15;
    }

    public static double toFahrenheit(double celsius) {
        return celsius * 9.0 / 5.0 + 32.0;
    }

    @Override
    public double getCelsius() {
        if (autoUpdateWhenPropertyRead)
            update();

        return celsius;
    }

    @Override
    public boolean isAutoUpdateWhenPropertyRead() {
        return autoUpdateWhenPropertyRead;
    }

    @Override
    public void setAutoUpdateWhenPropertyRead(boolean value) {
        autoUpdateWhenPropertyRead = value;
    }

    @Override
    public double getFahrenheit() {
        return TemperatureSensor.toFahrenheit(getCelsius());
    }

    @Override
    public double getKelvin() {
        return TemperatureSensor.toKelvin(getCelsius());
    }
}

File: Python/treehopper/libraries/sensors/temperature/temperature_sensor.py
from abc import abstractmethod
from treehopper.libraries.sensors.pollable import Pollable


class TemperatureSensor(Pollable):
    """Temperature sensor"""
    def __init__(self):
        super().__init__()
        self._celsius = 0

    @abstractmethod
    def update(self):
        """Request an update from the temperature sensor"""
        pass

    @property
    def celsius(self) -> float:
        """Gets the temperature, in degrees Celsius"""
        if self.auto_update_when_property_read:
            self.update()

        return self._celsius

    @property
    def fahrenheit(self):
        """Gets the temperature, in degrees Fahrenheit"""
        return self.celsius * 9.0 / 5.0 + 32.0

    @property
    def kelvin(self):
        """Gets the temperature, in Kelvin"""
        return self.celsius + 273.15
